src/pages/workspaces/workspace/notebooks/NotebookLauncher.js

```diff
@@ -0,0 +1,150 @@
+import _ from 'lodash/fp'
+import { Fragment } from 'react'
+import { div, h } from 'react-hyperscript-helpers'
+import { icon, spinner } from 'src/components/icons'
+import { TopBar } from 'src/components/TopBar'
+import { Rawls, Leo } from 'src/libs/ajax'
+import { getBasicProfile } from 'src/libs/auth'
+import { reportError } from 'src/libs/error'
+import * as Style from 'src/libs/style'
+import * as Nav from 'src/libs/nav'
+import * as Utils from 'src/libs/utils'
+import { Component } from 'src/libs/wrapped-components'
+
+
+const wait = ms => new Promise(resolve => setTimeout(resolve, ms))
+
+
+const styles = {
+  pageContainer: {
+    padding: '2rem'
+  },
+  step: {
+    container: {
+      display: 'flex',
+      alignItems: 'center',
+      lineHeight: '2rem',
+      margin: '0.5rem 0'
+    },
+    col1: {
+      flex: '0 0 30px'
+    },
+    col2: {
+      flex: 1
+    }
+  }
+}
+
+
+class NotebookLauncher extends Component {
+  constructor(props) {
+    super(props)
+
+    this.creator = getBasicProfile().getEmail()
+  }
+
+  async componentDidMount() {
+    const bucketName = await this.resolveBucketName()
+    this.setState({ bucketName })
+    const cluster = await this.startCluster()
+    await this.localizeNotebook(cluster)
+
+    const { name: workspaceName, notebookName } = this.props
+    window.location.href = `${cluster.clusterUrl}/notebooks/${workspaceName}/${notebookName}`
+  }
+
+  async resolveBucketName() {
+    const { namespace, name: workspaceName } = this.props
+
+    try {
+      const { workspace: { bucketName } } = await Rawls.workspace(namespace, workspaceName).details()
+      return bucketName
+    } catch (error) {
+      reportError('Error during Google bucket lookup', error)
+    }
+  }
+
+  async getCluster() {
+    const { namespace } = this.props
+
+    return _.flow(
+      _.filter({ googleProject: namespace, creator: this.creator }),
+      _.remove({ status: 'Deleting' }),
+      _.sortBy('createdDate'),
+      _.last
+    )(await Leo.clustersList())
+  }
+
+  async startCluster() {
+    while (true) {
+      const cluster = await this.getCluster()
+      const { status, googleProject, clusterName } = cluster
```

Comment:
There could be edge cases where the user doesn't have a cluster. We should make sure not to crash in that case.